Corona-Warn-App/src/main/java/de/rki/coronawarnapp/submission/ExposureKeyHistoryCalculations.kt

```diff
@@ -1,39 +1,73 @@
 package de.rki.coronawarnapp.submission
 
+import KeyExportFormat
+import androidx.annotation.VisibleForTesting
 import com.google.android.gms.nearby.exposurenotification.TemporaryExposureKey
+import de.rki.coronawarnapp.util.TimeAndDateExtensions
+import org.joda.time.DateTime
+import org.joda.time.DateTimeZone
+import org.joda.time.Instant
 
 class ExposureKeyHistoryCalculations(
     private val transmissionRiskVectorDeterminator: TransmissionRiskVectorDeterminator,
+    private val daysSinceOnsetOfSymptomsVectorDeterminator: DaysSinceOnsetOfSymptomsVectorDeterminator,
     private val keyConverter: KeyConverter
 ) {
 
+    companion object {
+        const val MAX_AGE_IN_DAYS = 14
+        const val TEN_MINUTES_IN_MILLIS = (10 * 60 * 1000).toLong()
+    }
+
     fun transformToKeyHistoryInExternalFormat(
         keys: List<TemporaryExposureKey>,
         symptoms: Symptoms
     ) =
         toExternalFormat(
-            toSortedHistory(limitKeyCount(keys)),
-            transmissionRiskVectorDeterminator.determine(symptoms)
+            removeOldKeys(sortWithRecentKeyFirst(keys)),
+            transmissionRiskVectorDeterminator.determine(symptoms),
+            daysSinceOnsetOfSymptomsVectorDeterminator.determine(symptoms)
         )
 
-    fun <T> limitKeyCount(keys: List<T>): List<T> =
-        keys.take(MAXIMUM_KEYS)
-
-    fun toExternalFormat(
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun removeOldKeys(
         keys: List<TemporaryExposureKey>,
-        transmissionRiskVector: TransmissionRiskVector
+        now: DateTime = Instant().toDateTime(DateTimeZone.UTC)
     ) =
-        keys.mapIndexed { index, key ->
-            // The latest key we receive is from yesterday (i.e. 1 day ago),
-            // thus we need use index+1
-            keyConverter.toExternalFormat(key, transmissionRiskVector.getRiskValue(index + 1))
+        keys.filter { it.ageInDays(now) in 0..MAX_AGE_IN_DAYS }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun toExternalFormat(
+        keys: List<TemporaryExposureKey>,
+        transmissionRiskVector: TransmissionRiskVector,
+        daysSinceOnsetOfSymptomsVector: DaysSinceOnsetOfSymptomsVector,
+        now: DateTime = Instant().toDateTime(DateTimeZone.UTC)
+    ): List<KeyExportFormat.TemporaryExposureKey> {
+        val result = mutableListOf<KeyExportFormat.TemporaryExposureKey>()
+        keys.groupBy { it.ageInDays(now) }.forEach { entry ->
+            val ageInDays = entry.key
+            entry.value.forEach {
```

Comment:
same her. for better readability it might be better to seperate the second forEach Block to an own method. but also just personal preference.

Reply:
I don't have any preference about this but if you decide to do this, I wouldn't see a need to change any of the related tests. It's just an internal refactoring of that function for improved readability.